File: cli/src/validation.rs
//! Validation utilities for the cutl CLI
//!
//! Validates URLs, short codes, and TTL formats.

use anyhow::{bail, Context};

/// Validates that a URL is well-formed and safe
///
/// # Rules
/// - Must start with `http://` or `https://`
/// - Cannot point to `localhost` or `127.0.0.1`
pub fn validate_url(url: &str) -> anyhow::Result<()> {
    // Check that URL starts with http:// or https://
    if !url.starts_with("http://") && !url.starts_with("https://") {
        bail!("URL must start with http:// or https://");
    }

    // Try to parse as URL to validate further
    let parsed = url::Url::parse(url).context("Invalid URL format")?;

    // Reject localhost and 127.0.0.1
    let host = parsed.host_str().unwrap_or("");
    if host == "localhost" || host.starts_with("127.0.0.1") {
        bail!("URL cannot point to localhost or 127.0.0.1");
    }

    Ok(())
}

/// Validates that a custom code matches the required pattern
///
/// # Rules
/// - Length: 1-32 characters
/// - Characters: only alphanumeric, hyphen, and underscore
pub fn validate_code(code: &str) -> anyhow::Result<()> {
    // Check length
    if code.is_empty() {
        bail!("Code cannot be empty");
    }
    if code.len() > 32 {
        bail!("Code cannot exceed 32 characters");
    }

    // Check characters: only alphanumeric, hyphen, and underscore
    if !code
        .chars()
        .all(|c| c.is_alphanumeric() || c == '-' || c == '_')
    {
        bail!("Code can only contain letters, numbers, hyphens, and underscores");
    }

    Ok(())
}

/// Validates the format of a TTL string
///
/// # Supported formats
/// - `5s` - 5 seconds
/// - `5m` - 5 minutes
/// - `1h` - 1 hour
/// - `1d` - 1 day
pub fn validate_ttl_format(ttl: &str) -> anyhow::Result<()> {
    let ttl = ttl.trim().to_lowercase();

    if ttl.len() < 2 {
        bail!("Invalid TTL format. Use format like 5m, 1h, 3d");
    }

    let (num_str, unit) = ttl.split_at(ttl.len() - 1);

    // Check that the number part is valid
    num_str.parse::<u64>().context("Invalid TTL number")?;

    // Check that the unit is valid
    match unit {
        "s" | "m" | "h" | "d" => Ok(()),
        _ => bail!("Invalid TTL unit: {}. Use s, m, h, or d", unit),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_validate_url_valid() {
        assert!(validate_url("https://example.com").is_ok());
        assert!(validate_url("http://example.com").is_ok());
    }

    #[test]
    fn test_validate_url_invalid() {
        assert!(validate_url("ftp://example.com").is_err());
        assert!(validate_url("localhost").is_err());
        assert!(validate_url("https://localhost").is_err());
        assert!(validate_url("https://127.0.0.1").is_err());
    }

    #[test]
    fn test_validate_code_valid() {
        assert!(validate_code("abc").is_ok());
        assert!(validate_code("ABC-123_test").is_ok());
        assert!(validate_code("a").is_ok());
        assert!(validate_code("a".repeat(32).as_str()).is_ok());
    }

    #[test]
    fn test_validate_code_invalid() {
        assert!(validate_code("").is_err());
        assert!(validate_code("a".repeat(33).as_str()).is_err());
        assert!(validate_code("abc@def").is_err());
        assert!(validate_code("abc def").is_err());
    }

    #[test]
    fn test_validate_ttl_format_valid() {
        assert!(validate_ttl_format("5s").is_ok());
        assert!(validate_ttl_format("5m").is_ok());
        assert!(validate_ttl_format("1h").is_ok());
        assert!(validate_ttl_format("3d").is_ok());
    }

    #[test]
    fn test_validate_ttl_format_invalid() {
        assert!(validate_ttl_format("5").is_err());
        assert!(validate_ttl_format("1w").is_err());
        assert!(validate_ttl_format("abc").is_err());
    }
}

File: frontend/src/utils/validation.ts
import type { ValidationError } from '../api/types';

/**
 * Validates a URL string
 * @param url - The URL to validate
 * @returns Error message if invalid, null if valid
 */
export function validateUrl(url: string): string | null {
  if (!url || url.trim() === '') {
    return 'URL is required';
  }

  try {
    const urlObj = new URL(url);

    // Ensure protocol is http or https
    if (urlObj.protocol !== 'http:' && urlObj.protocol !== 'https:') {
      return 'URL must use HTTP or HTTPS protocol';
    }

    return null;
  } catch {
    return 'Please enter a valid URL (e.g., https://example.com)';
  }
}

/**
 * Validates a custom code
 * @param code - The custom code to validate
 * @returns Error message if invalid, null if valid
 */
export function validateCode(code: string): string | null {
  if (!code || code.trim() === '') {
    return null; // Optional field
  }

  const trimmedCode = code.trim();

  if (trimmedCode.length < 1) {
    return 'Custom code must be at least 1 character';
  }

  if (trimmedCode.length > 32) {
    return 'Custom code must be 32 characters or less';
  }

  // Allow alphanumeric, hyphens, and underscores
  const validPattern = /^[a-zA-Z0-9_-]+$/;
  if (!validPattern.test(trimmedCode)) {
    return 'Custom code can only contain letters, numbers, hyphens, and underscores';
  }

  return null;
}

/**
 * Validates a TTL (time to live) string
 * @param ttl - The TTL string to validate (e.g., "7d", "1h", "30m")
 * @returns Error message if invalid, null if valid
 */
export function validateTtl(ttl: string): string | null {
  if (!ttl || ttl.trim() === '') {
    return null; // Optional field
  }

  const trimmedTtl = ttl.trim();
  const ttlPattern = /^(\d+)([smhd])$/;
  const match = trimmedTtl.match(ttlPattern);

  if (!match) {
    return 'TTL must be in format: number + unit (s/m/h/d). Examples: 5m, 1h, 3d, 30d';
  }

  const value = parseInt(match[1], 10);
  const unit = match[2];

  // Convert to minutes for validation
  const unitToMinutes: Record<string, number> = {
    's': 1 / 60,
    'm': 1,
    'h': 60,
    'd': 60 * 24,
  };

  const minutes = value * unitToMinutes[unit];

  // Minimum: 5 minutes
  if (minutes < 5) {
    return 'TTL must be at least 5 minutes';
  }

  // Maximum: 30 days
  const maxMinutes = 30 * 24 * 60;
  if (minutes > maxMinutes) {
    return 'TTL must be 30 days or less';
  }

  return null;
}

/**
 * Validates all form fields
 * @param url - The URL to validate
 * @param code - The custom code to validate
 * @param ttl - The TTL to validate
 * @returns Array of validation errors (empty if all valid)
 */
export function validateForm(
  url: string,
  code: string,
  ttl: string
): ValidationError[] {
  const errors: ValidationError[] = [];

  const urlError = validateUrl(url);
  if (urlError) {
    errors.push({ field: 'url', message: urlError });
  }

  const codeError = validateCode(code);
  if (codeError) {
    errors.push({ field: 'code', message: codeError });
  }

  const ttlError = validateTtl(ttl);
  if (ttlError) {
    errors.push({ field: 'ttl', message: ttlError });
  }

  return errors;
}
